File: public/js/midp/Display.js
/** javax.microedition.lcdui.Display / Displayable */
export class Displayable {
  constructor() {
    this._commands = [];
  }
}

export class Display {
  constructor(midlet) {
    this._midlet = midlet;
    this._current = null;
    this._dom = null;
  }

  static getDisplay(midlet) {
    if (!midlet._display) midlet._display = new Display(midlet);
    return midlet._display;
  }

  setCurrent(displayable) {
    this._current = displayable;
    if (displayable && typeof displayable.attach === 'function' && this._dom) {
      displayable.attach(this._dom, this);
      displayable.repaint();
    }
  }

  getCurrent() {
    return this._current;
  }

  /** Host page binds the HTML canvas once */
  bindDomCanvas(canvasEl) {
    this._dom = canvasEl;
    if (this._current && typeof this._current.attach === 'function') {
      this._current.attach(canvasEl, this);
      this._current.repaint();
    }
  }

  callSerially(r) {
    queueMicrotask(() => {
      try {
        r.run ? r.run() : r();
      } catch (e) {
        console.error(e);
      }
    });
  }

  vibrate(duration) {
    if (navigator.vibrate) navigator.vibrate(duration);
  }

  flashBacklight(_duration) {}
}

File: teavm-port/src/main/java/javax/microedition/lcdui/Display.java

package javax.microedition.lcdui;
import javax.microedition.midlet.MIDlet;
import bootstrap.JsBridge;
import org.teavm.jso.dom.html.HTMLCanvasElement;
import org.teavm.jso.browser.Window;

public class Display {
    private static Display instance;
    private Displayable current;
    private final MIDlet midlet;
    private HTMLCanvasElement dom;

    private Display(MIDlet m) {
        this.midlet = m;
        this.dom = JsBridge.gameCanvas();
    }

    public static void initForWeb(MIDlet m) {
        instance = new Display(m);
    }

    public static Display getDisplay(MIDlet m) {
        if (instance == null) instance = new Display(m);
        return instance;
    }

    public void setCurrent(Displayable next) {
        this.current = next;
        if (next instanceof Canvas) {
            Canvas c = (Canvas) next;
            if (dom != null) {
                c.attachDom(dom);
            }
            c.showNotify();
            c.repaint();
        }
    }

    public Displayable getCurrent() { return current; }

    public boolean flashBacklight(int duration) { return false; }
    public boolean vibrate(int duration) { return false; }
    public int numColors() { return 65536; }
    public boolean isColor() { return true; }
    public int numAlphaLevels() { return 256; }
    public void callSerially(Runnable r) {
        Window.requestAnimationFrame(ts -> r.run());
    }
}
